blockindex/contractstaking/cache.go

```diff
@@ -331,21 +358,23 @@ func (s *contractStakingCache) putBucketType(id uint64, bt *BucketType) {
 }
 
 func (s *contractStakingCache) putBucketInfo(id uint64, bi *bucketInfo) {
-	oldBi := s.bucketInfoMap[id]
+	// delete old candidate bucket map
+	if oldBi, existed := s.bucketInfoMap[id]; existed {
+		oldDelegate := oldBi.Delegate.String()
+		if _, existed := s.candidateBucketMap[oldDelegate]; existed {
+			delete(s.candidateBucketMap[oldDelegate], id)
+			if len(s.candidateBucketMap[oldDelegate]) == 0 {
+				delete(s.candidateBucketMap, oldDelegate)
+			}
+		}
+	}
 	s.bucketInfoMap[id] = bi
 	// update candidate bucket map
 	newDelegate := bi.Delegate.String()
 	if _, ok := s.candidateBucketMap[newDelegate]; !ok {
 		s.candidateBucketMap[newDelegate] = make(map[uint64]bool)
 	}
 	s.candidateBucketMap[newDelegate][id] = true
-	// delete old candidate bucket map
-	if oldBi != nil {
-		oldDelegate := oldBi.Delegate.String()
-		if oldDelegate != newDelegate {
-			delete(s.candidateBucketMap[oldDelegate], id)
-		}
```

Comment:
how likely is the case old bucket needs to be deleted? if it's not too much (like < 5-10%), then I think just add 2 lines (to delete empty map here) would suffice:
```
if len(s.candidateBucketMap[oldDelegate]) == 0 {
				delete(s.candidateBucketMap, oldDelegate)
			}
```

Reply:
the efficiency doesn't matter much here, otherwise more `if condition` will harm readability

Reply:
given the same correct logic, i would say we should aim for better code
the `if` here indicates the need to remove old/existing bucket, it's not too excessive `if` i think